tools: update checker

CvH opened the pull request:
# Tool to check for package updates upstream
* adds a package id to all packages to track the packages at https://release-monitoring.org/ (added already all missing packages), all we do is calling the api for the latest version
* adds a script that checks local version and compares it to upstream
* we (tx @antonlacon & @heitbaum ) test this for a while now and found no major prblems

# special github support 
if we use some random git hash from a github repo we can provide additional informations
needs github api token, fallback just shows local commit and upstream version if available 
* checks local commit we use for commit date
* shows available latest commit (with date) and latest tag (with date) to get an idea if update is needed

with github api
```
squeezelite                 | Commit date: 2020-11-16                  | Master: 2020-12-16 | TAG: no tags
tini                        | Commit date: 2018-04-21                  | Master: 2020-11-14 | TAG: v0.19.0 (2020-04-19)
```

without git api and if upstream is not github
```
squeezelite                 | cf2230bd8edb9397cafe97c94cda1836269ba7d9 | 1.9.8.1307
tini                        | fec3683b971d9c3ef73f284f176672c44b448662 | 0.19.0
```
# depends
* just bash, curl and jq

# usage
To check for all packages (ignores all packages we bump with dedicated scripts like kodi-addons)
~4min to complete.
`./tools/update-scan2`

To check a single package
`./tools/update-scan2 tini`

# full output

<details>
  <summary>click for full script output</summary>

```
christian@ubuntu:~/Documents/Git/LE-CvH$ ./tools/update-scan2
Github api usage activated

Updates found:

Package                             | LE git master                            | upstream location   
--------------------------------------------------------------------------------------------------------------------------------------------
MarkupSafe                          | 1.1.1                                    | 2.0.0a1             
Python3                             | 3.8.6                                    | 3.10.0a3            
SDL2                                | 2.0.10                                   | 2.0.14              
aixlog                              | 1.2.1                                    | 1.4.0               
aom                                 | 8c113ea88bc5afc1d5d7001c8dedbde7af2c82dc | 7                   
apache-ant                          | 1.10.8                                   | 1.10.9              
argtable2                           | 2.13                                     | 13                  
asio                                | 1.12.0                                   | 1.18.1              
at-spi2-atk                         | 2.34.2                                   | 2.38.0              
at-spi2-core                        | 2.36.0                                   | 2.38.0              
atf                                 | 2.2                                      | 2.4                 
autoconf                            | 2.69                                     | 2.70                
automake                            | 1.15.1                                   | 1.16.3              
avahi                               | 0.7                                      | 0.8                 
bash                                | 4.3.46                                   | 5.1                 
bchunk                              | 1.2.0                                    | 1.2.2               
bison                               | 3.5.4                                    | 3.7.4               
bitstream                           | 1.5                                      | 2.6.0.1             
bluez                               | 5.55                                     | 5.54                
boost                               | 1.74.0                                   | 1.75.0              
btrfs-progs                         | 4.8.4                                    | 5.9                 
bwm-ng                              | 0.6.1                                    | 0.6.2               
cairo                               | 1.16.0                                   | 1.17.4              
ccache                              | 3.7.7                                    | 4.1                 
ccid                                | 1.4.28                                   | 1.4.33              
cmake                               | 3.18.4                                   | 3.19.2              
connman                             | 40947be3848b0a864a366f8d67ce347265ea98c4 | 1.38                
containerd                          | 1.3.9                                    | 1.4.3               
crossguid                           | Commit date: 2015-08-03                  | Master: 2019-05-29 | TAG: v0.2.2 (2017-08-04)
cups                                | 2.2.8                                    | 2.3.3op1            
curl                                | 7.73.0                                   | 7.74.0              
cxxtools                            | 2.2.1                                    | 3.0                 
diskdev_cmds                        | 332.14                                   | 641.0.1             
docker                              | 19.03.14                                 | 20.10.1             
dotnet-runtime                      | 3.1.9                                    | 3.1.10              
dstat                               | Commit date: 2016-11-23                  | Master: 2020-06-18 | TAG: v0.7.4 (2019-01-17)
dvb-apps                            | 3d43b280298c                             | 1.1.1               
dvblast                             | 6fa5ef52443280d293e606510991917ddfbff705 | 3.4                 
ecm-tools                           | Commit date: 2015-01-21                  | Master: 2017-05-17 | TAG: v1.0.3 (2017-05-17)
efivar                              | Commit date: 2019-06-19                  | Master: 2020-10-15 | TAG: efivar-0.14 (2014-10-09)
elfutils                            | 0.178                                    | 0.182               
enet                                | Commit date: 2016-10-02                  | Master: 2020-06-13 | TAG: v1.3.15 (2020-04-21)
ethtool                             | 5.4                                      | 5.10                
expat                               | 2.2.9                                    | 2.2.10              
faad2                               | 2.8.8                                    | 2.10.0              
fakeroot                            | 1.24                                     | 1.25.3              
fdupes                              | 2.0.0                                    | 2.1.2               
ffmpegx                             | 4.3                                      | 4.3.1               
ffmpeg                              | 4.3.1-Matrix-Beta1                       | 4.3.1               
flashrom                            | 1.2                                      | 1.1                 
flex                                | 2.6.0                                    | 2.6.4               
fluidsynth                          | 1.1.6                                    | 2.1.5               
fontconfig                          | 2.13.1                                   | 2.13.93             
fuse                                | 2.9.9                                    | 2.23                
gcc-arm-aarch64-none-elf            | 9.2-2019.12                              | 10.2-2020.11        
gcc-arm-aarch64-none-linux-gnu      | 9.2-2019.12                              | 10.2-2020.11        
gcc-arm-arm-none-linux-gnueabihf    | 9.2-2019.12                              | 10.2-2020.11        
gdk-pixbuf                          | 2.38.0                                   | 2.42.2              
gettext                             | 0.20.2                                   | 0.21                
glfw                                | 2.7.9                                    | 3.3.2               
glmark2                             | Commit date: 2019-07-08                  | Master: 2020-11-14 | TAG: 2020.04 (2020-04-28)
gmp                                 | 6.2.0                                    | 6.2.1               
gnutls                              | 3.6.15                                   | 3.7.0               
go                                  | 1.14.12                                  | 1.15.6              
gpiozero                            | 1.5.0                                    | 1.5.1               
gptfdisk                            | 1.0.4                                    | 1.0.5               
grub                                | 6a34fdb76a07305b95e31659bc27b1d190101cbf | 0.0.10              
gtk3                                | 3.24.23                                  | 3.94.0              
harfbuzz                            | 1.8.1                                    | 2.7.4               
hdparm                              | 9.58                                     | 9.60                
htop                                | 3.0.0beta5                               |                     
i2c-tools                           | 3.1.2                                    | 4.2                 
icu                                 | 68.1                                     | 68-2                
inadyn                              | 2.5                                      | 2.7                 
inotify-tools                       | 3.20.1                                   | 3.20.11.0           
intel-ucode                         | 20201118                                 | 20190918            
iperf                               | 3.6                                      | 3.9                 
iptables                            | 1.8.4                                    | 1.8.6               
irssi                               | 1.1.2                                    | 1.2.2               
iwd                                 | 1.5                                      | 1.10                
iw                                  | 5.4                                      | 5.9                 
jasper                              | 1.900.1                                  | 2.0.23              
jsoncpp                             | 1.9.2                                    | 1.9.4               
kernel-firmware                     | 20201022                                 | 20201218            
keyutils                            | 1.6                                      | 1.6.3               
kmod                                | 24                                       | 27                  
lame                                | 3.100                                    | 398                 
lcms2                               | 2.9                                      | 2.11                
lftp                                | 4.8.4                                    | 4.9.2               
libX11                              | 1.6.12                                   | 1.7.0               
libXScrnSaver                       | 1.2.2                                    | 1.2.3               
libaacs                             | 0.9.0                                    | 0.11.0              
libarchive                          | 3.5.0                                    | 3.5.1               
libass                              | 0.14.0                                   | 0.15.0              
libbluray                           | 1.1.2                                    | 1.2.1               
libcap                              | 2.45                                     | 2.46                
libcec                              | 4.0.7                                    | 6.0.2               
libconfig                           | 1.5                                      | 1.7.2               
libconfuse                          | 3.2.2                                    | 3.3                 
libde265                            | Commit date: 2019-08-16                  | Master: 2020-11-02 | TAG: v1.0.8 (2020-11-02)
libdrm                              | 2.4.102                                  | 2.4.103             
libdvbcsa                           | Commit date: 2017-02-06                  | Master: 2018-01-28 | TAG: no tags
libdvdcss                           | 1.4.2-Leia-Beta-5                        | 1.4.2               
libdvdnav                           | Commit date: 2018-05-03                  | Master: 2017-02-18 | TAG: 6.0.0 (2018-01-18)
libdvdread                          | Commit date: 2018-05-03                  | Master: 2016-02-20 | TAG: 6.0.0 (2018-01-14)
libepoxy                            | 1.5.4                                    | 1.5.5               
liberation-fonts-ttf                | 2.00.1                                   | 2.1.2               
libestr                             | 0.1.10                                   | 0.1.11              
libevdev                            | 1.8.0                                    | 1.10.0              
libexif                             | 0.6.21                                   | 0.6.22              
libfmt                              | 7.0.3                                    | 7.1.3               
libftdi1                            | 1.4                                      | 1.5                 
libgcrypt                           | 1.8.6                                    | 1.8.7               
libgpg-error                        | 1.39                                     | 1.41                
libhdhomerun                        | Commit date: 2019-06-25                  | Master: 2020-03-03 | TAG: no tags
libheif                             | 1.6.2                                    | 1.10.0              
libiconv                            | 1.15                                     | 1.16                
libjpeg-turbo                       | 2.0.5                                    | 2.0.90              
liblognorm                          | 2.0.5                                    | 2.0.6               
libmediainfo                        | 18.05                                    | 20.09               
libmicrohttpd                       | 0.9.71                                   | 0.9.72              
libmtp                              | 1.1.13                                   | 1.1.18              
libnetwork                          | Commit date: 2020-11-24                  | Master: 2020-12-15 | TAG: v0.8.0-dev.2 (2016-05-07)
libnftnl                            | 1.1.5                                    | 1.1.8               
libnl                               | 3.4.0                                    | 3_5_0               
libopenmpt                          | 0.4.11                                   | 0.5.4+release.msvc  
libpcap                             | 1.7.4                                    | 1.9.1               
libpng                              | 1.6.37                                   | 1.7.0beta89         
libprojectM                         | 3.1.2                                    | 3.1.8               
libraw                              | 0.19.5                                   | 0.20.2              
librelp                             | 1.2.17                                   | 1.9.0               
libssh2                             | 1.8.0                                    | 1.9.0               
libssh                              | 0.9.3                                    | 0.9.5               
libstatgrab                         | 0.91                                     | 0.92                
libtirpc                            | 1.2.6                                    | 1.3.1               
libudfread                          | 1.1.0                                    | 1.1.1               
libugpio                            | 0.0.6                                    | 0.0.4               
libusb-compat                       | 0.1.5                                    | 1.0.18              
libusb                              | 1.0.23                                   | 2.0.2               
libva-utils                         | 2.9.1                                    | 2.10.0              
libvncserver                        | 0.9.11                                   | 0.9.13              
libvpx                              | 1.8.2                                    | 1.9.0               
libwebsockets                       | 4.0.7                                    | 4.1.6               
libxss                              | 1.2.2                                    | 1.2.3               
libzen                              | 0.4.37                                   | 0.4.38              
libzip                              | 1.6.1                                    | 1.7.3               
linuxconsoletools                   | 1.5.1                                    | 1.7.0               
llvm                                | 10.0.1                                   | 12-init             
lm_sensors                          | Commit date: 2018-11-23                  | Master: 2020-12-02 | TAG: i2c-2-8-km2 (2003-01-21)
mariadb                             | 10.4.17                                  | 10.5.8              
mediainfo                           | 18.05                                    | 20.09               
mesa-demos                          | 8.3.0                                    | 8.4.0               
minidlna                            | 799e6cf505ec470b2bf0ae4118143380aa16b837 | 1.3.0               
minisatip                           | 1.0.3                                    | 1.0.4               
miniupnpc                           | 2.1                                      | 2.2.1               
minizip                             | 2.9.2                                    | 2.10.5              
mpc                                 | 1.1.0                                    | 1.2.1               
mpfr                                | 4.0.2                                    | 4.1.0               
mpg123                              | 1.25.10                                  | 1.26.4              
mpv-drmprime                        | 0.30.0                                   | 0.33.0              
mtools                              | 4.0.24                                   | 4.0.26              
nano                                | 4.9.3                                    | 5.4                 
ncurses                             | 6.1-20181215                             | 6.2-20201227        
net-snmp                            | 5.8                                      | 6.0.1               
netbase                             | 6.1                                      | 6.2                 
nghttp2                             | 1.41.0                                   | 1.42.0              
ngrep                               | Commit date: 2018-01-01                  | Master: 2019-01-28 | TAG: V1_47 (2017-09-07)
nlohmann-json                       | 3.7.3                                    | 3.9.1               
nmap                                | 7.70                                     | 7.91                
nspr                                | 4.19                                     | 3.60                
nss                                 | 3.47.1                                   | 3.60                
ntfs-3g_ntfsprogs                   | 2017.3.23AR.5                            | 2017.3.23           
oniguruma                           | 6.9.1                                    | 6.9.6               
open-iscsi                          | Commit date: 2013-07-29                  | Master: 2020-12-18 | TAG: 2.1.3 (2020-12-18)
open-vm-tools                       | 11.1.0                                   | 11.2.0              
openal-soft                         | 1.20.1                                   | 1.21.0              
opencaster                          | 3.2.2                                    | 3.2.2+dfsg          
openvpn                             | 2.4.10                                   | 2.5.0               
pango                               | 1.43.0                                   | 1.48.0              
pciutils                            | 3.6.4                                    | 3.7.0               
pcre                                | 8.43                                     | 8.44                
pcsc-lite                           | 1.8.23                                   | 1.9.0               
pixman                              | 0.38.4                                   | 0.40.0              
pngquant                            | 2.12.0                                   | 2.13.1              
popt                                | 1.16                                     | 1.18                
populatefs                          | 1.0                                      | 0.9                 
proftpd                             | 1.3.6                                    | 1.3.7a              
protobuf                            | 3.1.0                                    | 4.0.0-rc2           
pugixml                             | 1.10                                     | 1.11.4              
pulseaudio                          | 13.0                                     | 14.0                
pycryptodome                        | 3.9.8                                    | 3.9.9               
pygobject                           | 2.28.7                                   | 3.38.0              
qemu                                | 4.0.0                                    | 5.2.0               
qtbase                              | 5.14.0                                   | 6.0.0               
rar2fs                              | 1.27.1                                   | 1.29.3              
readline                            | 8.0                                      | 8.1                 
rpcbind                             | 0.2.4                                    | 1.2.5               
rsyslog                             | 8.37.0                                   | 8.2012.0            
rustup.rs                           | 1.22.1                                   | 1.23.1              
rust                                | 1.45.0                                   | 1.48.0              
samba                               | 4.13.3                                   | 4.13.0rc4           
sbc                                 | 1.4                                      | 1.5                 
serdisplib                          | 1.97.9                                   | 2.02                
setuptools                          | 51.0.0                                   | 51.1.0.post20201221 
shairport-sync                      | 3.3.6                                    | 3.3.7d12            
shared-mime-info                    | 1.9                                      | 2.0                 
slang                               | 2.1.4                                    | 2.3.2               
smartmontools                       | 7.0                                      | 7.1                 
snapcast                            | 0.15.0                                   | 0.22.0              
spdlog                              | 1.8.1                                    | 1.8.2               
sqlite                              | 3.3.4                                    | 3.34.0              
squeezelite                         | Commit date: 2020-11-16                  | Master: 2020-12-16 | TAG: no tags
sshfs                               | 2.10                                     | 3.7.1               
strace                              | 5.9                                      | 5.10                
stress-ng                           | 0.11.14                                  | 0.12.01             
swig                                | 4.0.1                                    | 4.0.2               
syncthing                           | 1.5.0                                    | 1.12.1-rc.1         
systemd                             | 247                                      | 247.2               
taglib                              | 1.11.1                                   | 1.12-beta-1         
tcpdump                             | 4.9.2                                    | 4.9.3               
terminus-font                       | 4.48                                     | 4.49.1              
tiff                                | 4.1.0                                    | 4.2.0               
tigervnc                            | 1.9.0                                    | 1.11.0              
tinc                                | 1.1pre17                                 | 1.1pre14            
tini                                | Commit date: 2018-04-21                  | Master: 2020-11-14 | TAG: v0.19.0 (2020-04-19)
tinyxml2                            | 7.1.0                                    | 8.0.0               
tntnet                              | 2.2.1                                    | 3.0                 
tslib                               | 1.1                                      | 1.22                
ttyd                                | 1.6.1                                    | 1.6.2               
tvheadend42                         | Commit date: 2019-11-15                  | Master: 2020-12-16 | TAG: v4.3 (2017-04-24)
u-boot-tools                        | 2019.01                                  | 2020.10             
u-boot                              | Commit date: 2018-09-12                  | Master: 2017-10-17 | TAG: v2018.05-rc1 (2018-04-03)
udpxy                               | 1.0.23-12                                | 1.0-25.1            
unzip                               | 60                                       | 6.0                 
usb-modeswitch                      | 2.5.2                                    | 2.6.0               
usbutils                            | 008                                      | 013                 
vdpauinfo                           | 1.0                                      | 1.4                 
vdr-plugin-vnsiserver               | Commit date: 2018-10-19                  | Master: 2020-05-23 | TAG: v1.8.0 (2018-10-19)
vdr                                 | 2.4.4                                    | 2.4.6               
vim                                 | 8.2.0023                                 | 8.2.2232            
vsxu                                | 0.5.1                                    | 0.2.0               
wayland                             | 1.18.0                                   | 1.20                
whois                               | 5.2.13                                   | 5.5.7               
wireless-regdb                      | 2020.04.29                               | 2020.11.20          
x264                                | d4099dd4c722f52c4f3c14575d7d39eb8fadb97f | 20191217-2245       
xf86-input-libinput                 | 0.29.0                                   | 0.30.0              
xf86-video-intel                    | 9236c58252511d3e985cbd4c022a202c36314e3c | 2.99.917            
xf86-video-nvidia-legacy            | 340.108                                  | 340.102             
xf86-video-nvidia                   | 455.45.01                                | 460.27              
xkbcomp                             | 1.4.3                                    | 1.4.4               
xkeyboard-config                    | 2.29                                     | 2.31                
xmlstarlet                          | 1.6.1                                    | 1.6.7               


Current 213:
Mako Pillow RPi.GPIO alsa-lib alsa-plugins alsa-topology-conf alsa-ucm-conf alsa-utils atk attr autoconf-archive autossh binutils bkeymaps brcmfmac_sdio-firmware-rpi brcmfmac_sdio-firmware busybox bzip2 dav1d dbus-glib dbus-python dbussy dbus diffutils distutilscross dosfstools dtach dtc dvbsnoop e2fsprogs efibootmgr enca encfs encodings evtest file flac flatbuffers fluxbox font-bitstream-type1 font-cursor-misc font-misc-misc font-util font-xfree86-type1 freetype fribidi gcc gdb giflib glew glibc glib glm glu gmmlib hd-idle hddtemp heimdal hid_mapper iftop intel-vaapi-driver intltool jq json-c libICE libSM libXau libXcomposite libXcursor libXdamage libXext libXfixes libXfont2 libXft libXinerama libXi libXmu libXrandr libXrender libXtst libXt libXxf86vm libaio libbdplus libcdio libdaemon libdnet libev libfastjson libffi libfontenc libfstrcmp libglvnd libid3tag libidn2 libimobiledevice libinput libirman liblogging libmad libmnl libmodplug libmpdclient libmpeg2 libnfs libogg libpciaccess libplist libpthread-stubs libsamplerate libshairplay libsndfile libtool libunwind libusbmuxd libuv libva libvdpau libvorbis libxcb libxkbcommon libxkbfile libxml2 libxshmfence libxslt lirc lshw lz4 lzo m4 make mariadb-connector-c mc media-driver memtester mesa meson mpd-mpc mpd nasm ncftp nettle newt ninja nmon nss-mdns openssh openssl opus p7zip parted patch pathlib pigz pkg-config popl procps-ng pv rapidjson rpi-eeprom rsync runc screen scrnsaverproto sed setxkbmap sidplay-libs simplejson soxr speexdsp speex squashfs-tools syslinux tinyxml tsdecrypt tz udevil unclutter unrar usbmuxd util-linux util-macros v4l-utils valgrind vdr-plugin-streamdev w_scan waf wavpack wayland-protocols waylandpp weston wireguard-tools wireless_tools wpa_supplicant x265 xcb-proto xf86-input-evdev xf86-input-synaptics xf86-video-amdgpu xf86-video-ati xf86-video-vmware xorg-intel-gpu-tools xorg-launch-helper xorg-server xorgproto xrandr xtrans xz yajl zip zlib zstd zvbi 


No ID 119:
RTL8188EU RTL8192CU RTL8192DU RTL8192EU RTL8812AU amlogic-boot-fip arm-mem bcm2835-bootloader bcm2835-driver bcm_sta bcmstat blindscan-s2 bluetooth-audio boblightd brcmfmac_sdio-firmware-imx chrome chromium comskip configtools crazycat digital_devices dispmanx_vnc driverselect dvb-firmware dvb-latest dvb-tools edid-decode emby4 emby entropy estouchy ethmactool eventlircd evrepeat exynos-boot-fip ffmpeg-tools firmware-dragonboard firmware-imx ftgl getscancodes gnulib hdhomerun hyperion iat imon-mce installer ir-bpf-decoders iwlwifi-firmware jdk-aarch64-zulu jdk-arm-zulu jdk-x86_64-zulu jre.zulu kmscube lan951x-led-ctl lcdd libhid libmali librespot mali-bifrost mali-midgard mali-utgard media_tree_cc media_tree misc-firmware mkbootimg mono moonlight mtpfs multimedia-tools mumudvb network-tools oscam pcscd plymouth-lite remote repository.linuxserver.docker rkbin rpi-cirrus-config rpi-tools rpi_ws281x rtl8723bs_bt rtmpdump sapphire script.config.vdr snapclient snapserver soil steamcontroller st sundtek-mediatv system-tools szap-s2 t2scan texturecache.py touchscreen tstools tune-s2 tvh-dtv-scan-tables tvheadend tvmosaic u-boot-script vdr-addon vdr-plugin-ddci2 vdr-plugin-dummydevice vdr-plugin-dvbapi vdr-plugin-eepg vdr-plugin-epgfixer vdr-plugin-epgsearch vdr-plugin-iptv vdr-plugin-live vdr-plugin-restfulapi vdr-plugin-robotv vdr-plugin-satip vdr-plugin-wirbelscancontrol vdr-plugin-wirbelscan vdr-plugin-xmltv2vdr wait-time-sync webgrabplus wlan-firmware 

Ignored 30:
adafruit-libraries alsa autotools chrome-libXcomposite chrome-libXdamage chrome-libXfixes chrome-libXi chrome-libXrender chrome-libXtst chrome-libxcb chrome-libxkbcommon corefonts debug game-tools image initramfs jre-libXinerama jre-libbluray libc linux-drivers linux-firmware locale mediacenter misc-packages network oem testing toolchain virtual x11 
```
</details>

# todo
* squash commits
* change name `update-scan2` -> `update-scan`

# future
* the script could be run via github actions
* output could be piped to slack
* bash could be improved
* caching - already done by @heitbaum (not merged, just starting now less complex)

lrusak commented:
We already have distro-tool to check for package updates, do we need another one? Why not update distro-tool to use the release monitoring api?

CvH commented:
Goal was to create something "simple" and maintainable that actually works. 
distro-tool with 1100 loc bash with inline python is far away from simple and maintainable nor does it even work for me at all (nothing besides basic errors).

Ofc this could be glued ontop of distro-tool.

Current script from pr has ~15 loc logic and anything else is just for fancy and easy readable output.
The main call is 3 lines basic bash.

If you volunteer to include it at distro-tool it you are welcome. Personally I have no idea how to make it easy maintain and extendable.

lrusak commented:
I agree that `distro-tool` is a bit terse, however, it does it's job well.

Part of the problem is that it was built originally to keep others from using it. The other part of the problem is that it has two functions:
1) checking for new versions
2) syncing packages with our mirror

I think with a bit of effort `distro-tool` can be more user friendly and more valuable to the project.

If you have a problem using it all you have to do is ask 😃 The original author (@MilhouseVH) might even be willing to answer your questions 😉 

Sample output:
```
./tools/distro-tool --check-main -l /home/lukas/git/libreelec -t /home/lukas/git/libreelec/sources -s -d /home/lukas/git/libreelec/download --dry-run --all 


Synchronising LibreELEC.tv (branch: intel-project, version: devel) with SOURCE server /home/lukas/git/libreelec/sources/devel

Distro Source is: http://sources.libreelec.tv/devel
Distro Mirror is: http://sources.libreelec.tv/mirror
 Syncing against: /home/lukas/git/libreelec/sources/devel
  Downloading to: /home/lukas/git/libreelec/download/devel
     Check Newer: Yes  Class/Type: Main / All
         Dry run: Yes

New Ver Pkg available: Python3 (3.8.5 => 3.9.0) http://www.python.org/ftp/python/3.9.0/Python-3.9.0.tar.xz
New Ver Pkg available: Mako (1.1.1 => 1.1.2) https://files.pythonhosted.org/packages/source/M/Mako/Mako-1.1.2.tar.gz
New Ver Pkg available: Pillow (7.2.0 => 8.0.0) https://github.com/python-pillow/Pillow/archive/8.0.0.tar.gz
New Rev Pkg available: amlogic-boot-fip (c89bc11d2904e819baed20c7ddba5b5b42a09141 => a7b126764c0138579ba5eb74641f960fb67ab706) https://github.com/LibreELEC/amlogic-boot-fip/archive/a7b126764c0138579ba5eb74641f960fb67ab706.tar.gz
New Rev Pkg available: RTL8812AU (e9fbf5c051453941bbc029810b893a6c010714e6 => ad351bd0afeb47d2fb197aff8cde4d7fb5fead9e) https://github.com/aircrack-ng/rtl8812au/archive/ad351bd0afeb47d2fb197aff8cde4d7fb5fead9e.tar.gz
New Rev Pkg available: LibreELEC-settings (db288855d6a971ef29567ebc795301907e145797 => 3cf8fcdd1f7ddaff2a39f98201737454e458b8d9) https://github.com/LibreELEC/service.libreelec.settings/archive/3cf8fcdd1f7ddaff2a39f98201737454e458b8d9.tar.gz
New Rev Pkg available: RTL8192EU (1628c748ced6c18441440fb40bbc86d1a4aa1627 => b7faffdd77767269770b79876f88dd1145b6a630) https://github.com/Mange/rtl8192eu-linux-driver/archive/b7faffdd77767269770b79876f88dd1145b6a630.tar.gz
New Rev Pkg available: RTL8192CU (0f37bd7f97573f835e4dedb4b2628c1ceedd71ad => 7a15cb42e568b0f9c3c872bd6a06cdcab9f0ffd0) https://github.com/pvaret/rtl8192cu-fixes/archive/7a15cb42e568b0f9c3c872bd6a06cdcab9f0ffd0.tar.gz
New Rev Pkg available: RTL8188EU (c9280272bcc358e52d57b3f88e42596795b8f6c1 => b02c92bc92e22fd8b51c9acd6d602637cfce7256) https://github.com/lwfinger/rtl8188eu/archive/b02c92bc92e22fd8b51c9acd6d602637cfce7256.tar.gz
...
```
Full log: https://termbin.com/5o18 (looks bad because of color codes and threaded output)

heitbaum commented:
In looking at distro-tool it does have the enhanced scope for end-to-end whilst the proposed update-scan is focussed on checking exclusively using the upstream release-monitoring api. Adding the PKG_ID to all the packages would improve the capability of both scripts to report on / update to upstream. Tacking update-scan as a function to distro-tool to keep a single script would be a possibility, though would think that 2 sub functions in the same script would not be the right thing. Suggest that once the PKG_IDs are in place in the package.mk files and updating of distro for the release, then as a backlog item look at merging the tools?

lrusak commented:
> In looking at distro-tool it does have the enhanced scope for end-to-end whilst the proposed update-scan is focussed on checking exclusively using the upstream release-monitoring api. 

This isn't true as the update-scan script contains logic to scrap github using the github api. The disro-tool does this already and does it threaded.

>Adding the PKG_ID to all the packages would improve the capability of both scripts to report on / update to upstream.

I'm not really a fan of adding some random `PKG_ID` to each package. Many packages have empty ID's and it seems unnecessary anyways.

Tacking update-scan as a function to distro-tool to keep a single script would be a possibility, though would think that 2 sub functions in the same script would not be the right thing. Suggest that once the PKG_IDs are in place in the package.mk files and updating of distro for the release, then as a backlog item look at merging the tools?

I disagree. There is no rush on this so I would suggest doing it right in the first place. Adding more cruft is not a solution.


@CvH you might want to squash your commits. You've left a personal github token in one of your commits https://github.com/LibreELEC/LibreELEC.tv/pull/4809/commits/e42559f00aec10aafc66c1503712456d9aed8c08#diff-b4f61cdce85d5343785e961d7e4f40d8d1679331b007346afba09d4e7b19e9e8R10

dhewg commented:
Using release-monitoring.org is a good idea, way better than what distro-tool's check_newer() is doing atm.

But instead of adding PKG_ID everywhere, it looks like the proper way it to register a new distro over there and then add packages under that distro (which connects the release-monitoring.org project name with the LE package name), 

After that it's just `GET /api/project/(distro)/(path: package_name)` with LE's package name without any further mapping.

heitbaum commented:
That looks good. Based on the work done, should be a simple matter of running the api `POST /api/v2/packages/` https://release-monitoring.org/static/docs/api.html to load the data into release-monitoring.org

```
# curl -sL https://release-monitoring.org/api/project/1 | jq -r '.id, .ecosystem, .name'
1
http://sourceforge.net/projects/zero-install
0install
```

**`POST /api/v2/packages/`**
```
POST /api/v2/packages/ HTTP/1.1
Accept: application/json
Accept-Encoding: gzip, deflate
Authorization: Token gAOFi2wQPzUJFIfDkscAKjbJfXELCz0r44m57Ur2
Connection: keep-alive
Content-Length: 120
Content-Type: application/json
Host: localhost:5000
User-Agent: HTTPie/0.9.4

{
    "distribution": "LibreELEC",
    "package_name": "PKG_NAME",
    "project_ecosystem": ".ecosystem",
    "project_name": ".name"
}
```

heitbaum commented:
Just expermented with **bash** looks good.

```
LibreELEC:~ # curl -sL **https://release-monitoring.org/api/project/LibreELEC/bash**
{"backend":"GNU project","created_on":1412174950.0,"ecosystem":"http://www.gnu.org/software/bash/","homepage":"http://www.gnu.org/software/bash/","id":166,"name":"bash","packages":[{"distro":"Fedora","package_name":"bash"},{"distro":"PLD-Linux","package_name":"bash"},{"distro":"Alpine","package_name":"bash"},{"distro":"Arch Linux","package_name":"bash"},{"distro":"Cygwin","package_name":"bash"},{"distro":"OpenMandriva","package_name":"bash"},{"distro":"openSUSE","package_name":"bash"},{"distro":"Buildroot","package_name":"bash"},{"distro":"Timesys","package_name":"bash"},{"distro":"LibreELEC","package_name":"bash"}],"regex":null,"updated_on":1609232235.0,"version":"5.1","version_url":null,"versions":["5.1","5.0","4.4.18","4.4.12","4.4","4.3.30","4.3","4.2.53","4.2","4.1","4.0","3.2.57","3.2.48","3.2","3.1","3.0.16","3.0","2.05b","2.05a","2.05","2.04","2.03","2.02.1","2.02","2.01.1","2.01","2.0","1.14.7"]}
LibreELEC:~ # curl -sL https://release-monitoring.org/api/project/166
{"backend":"GNU project","created_on":1412174950.0,"ecosystem":"http://www.gnu.org/software/bash/","homepage":"http://www.gnu.org/software/bash/","id":166,"name":"bash","packages":[{"distro":"Fedora","package_name":"bash"},{"distro":"PLD-Linux","package_name":"bash"},{"distro":"Alpine","package_name":"bash"},{"distro":"Arch Linux","package_name":"bash"},{"distro":"Cygwin","package_name":"bash"},{"distro":"OpenMandriva","package_name":"bash"},{"distro":"openSUSE","package_name":"bash"},{"distro":"Buildroot","package_name":"bash"},{"distro":"Timesys","package_name":"bash"},{"distro":"LibreELEC","package_name":"bash"}],"regex":null,"updated_on":1609232235.0,"version":"5.1","version_url":null,"versions":["5.1","5.0","4.4.18","4.4.12","4.4","4.3.30","4.3","4.2.53","4.2","4.1","4.0","3.2.57","3.2.48","3.2","3.1","3.0.16","3.0","2.05b","2.05a","2.05","2.04","2.03","2.02.1","2.02","2.01.1","2.01","2.0","1.14.7"]}
LibreELEC:~ #
```

heitbaum commented:
JSON load file http://paste.ubuntu.com/p/ChxSNTKsB8/ ready - just reviewing before executing.

CvH commented:
@lrusak 
> I think with a bit of effort distro-tool can be more user friendly and more valuable to the project.

There are still major problems left after a cleanup
* we tried the distro-tool approach with uscan (debian version of distro-tool), resulting in ~20 loc and similar results (based on what distro-tool is doing). You have a lot packages not matching due obscure versioning etc that need babysitting and workarounds to make it work, with possible changes at any time that need manually adjustments (130ish packages at LE)
* uscan/distrotool checks just the url we provide, that is broken by design because you will never see new packages if the project shifts to some other location, fork ... whatnot and needs manual babysitting
* you have a output you can't trust, because you still need to manual check upstream if they changed anything that did not align with LE
* nobody is doing it, nobody babysits the packages, nobody maintains the workarounds, nobody checks manually upstream regularly, nobody maintains the complex distro-tool

Ofc the tracker we are using is not failure free, but there are much more ppl maintaining it and all logic/work is done there instead of us by reinventing the wheel.

@dhewg nice, didn't realised we can add a distro there, so we can drop the local mapping and just push it to the tracker mapping

@heitbaum looks good, just push the button - I adjust the script

heitbaum commented:
- @CvH - full stream ahead. 
- Data Checked. 
- Removed a couple that will need manual review (will cross reference using the new code logic against http://paste.ubuntu.com/p/dJ9P2tfknC/ once fully completed the upload of data.)
- Did a few others manually.

first block done (up to ntfs)
- http://paste.ubuntu.com/p/vQPq7Tx3Vf/

```
{
  "distribution": "LibreELEC",
  "package_name": "Mako",
  "project_ecosystem": "pypi",
  "project_name": "Mako"
}
```

```
#LibreELEC: 3915 Mako
HTTP/2 201 
date: Tue, 29 Dec 2020 12:42:52 GMT
server: Apache/2.4.43 (Fedora) mod_wsgi/4.6.8 Python/3.8
strict-transport-security: max-age=31536000; includeSubDomains; preload
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
referrer-policy: same-origin
content-length: 46
content-type: application/json
set-cookie: c7f2af7958ac7bdd3fe2e687de986bfb=7d2be215a66b17431bdeff9887db6c9b; path=/; HttpOnly; Secure
apptime: D=771430
x-fedora-proxyserver: proxy01.iad2.fedoraproject.org
x-fedora-requestid: X@skTLjoC8-Xvqd2jMQ58wABWAI

{"distribution": "LibreELEC", "name": "Mako"}
```

lrusak commented:
> @lrusak
> 
> > I think with a bit of effort distro-tool can be more user friendly and more valuable to the project.
> 
> There are still major problems left after a cleanup
> 
> * we tried the distro-tool approach with uscan (debian version of distro-tool), resulting in ~20 loc and similar results (based on what distro-tool is doing). You have a lot packages not matching due obscure versioning etc that need babysitting and workarounds to make it work, with possible changes at any time that need manually adjustments (130ish packages at LE)

How many of these packages actually need updating regularly?

> * uscan/distrotool checks just the url we provide, that is broken by design because you will never see new packages if the project shifts to some other location, fork ... whatnot and needs manual babysitting
> * you have a output you can't trust, because you still need to manual check upstream if they changed anything that did not align with LE
> * nobody is doing it, nobody babysits the packages, nobody maintains the workarounds, nobody checks manually upstream regularly, nobody maintains the complex distro-tool

It worked fine for us for that last few years 🤷 

> Ofc the tracker we are using is not failure free, but there are much more ppl maintaining it and all logic/work is done there instead of us by reinventing the wheel.

And now the plan is to just shift the logic to something that you have no idea how it checks in the backend and you just want to trust the output it provides? What happens when we shift to a different fork or version of a package? Where does that get updated and checked? (distro-tool will handle this because it's all local)

> @dhewg nice, didn't realised we can add a distro there, so we can drop the local mapping and just push it to the tracker mapping
> 

If you can make your update script standalone and not touch any packages then I'm fine adding it but it's still going to require someone to maintain it and add new packages to release-monitoring any time a new package is added to LibreELEC 🤷 

----

There is no one good solution here really. Having multiple solutions is fine I guess. Just don't discount what other people have done already and is proved to be working 👍 

----

As an aside, I assume because no one else knows how distro-tool works the package mirror hasn't been synced for a while?

heitbaum commented:
👍 Script working well against https://release-monitoring.org/
http://paste.ubuntu.com/p/2znZjG9ttF/

CvH commented:
- updated for the id/pkg_name handling at the release tracker -> dropped the PKG_ID change
- added exclusion for meta/virtual packages 
we have now basically complete coverage (likely some packages still wrong)

@lrusak 
> How many of these packages actually need updating regularly?

In a 1-2 year scale basically ~3/4 of all packages. Nobody checks them upstream manually for changes.

> It worked fine for us for that last few years 🤷

uscan does the same with ~20loc and we dropped the idea because the results are too bad and it needs far too much babysitting to get a complete result. It does not work for 1/3 of the packages due crappy filenames, versions and whatnot. And it does still not work if we provide a wrong urls and the other points pointed out.

> And now the plan is to just shift the logic to something that you have no idea how it checks in the backend and you just want to trust the output it provides?

Its basically like uscan but the exceptions and quirks are maintained at some website instead locally (nobody does it from us locally). 

> distro-tool will handle this because it's all local

if someone adds the needed quirks for all packages and maintain and checks them regularly - hint, nobody does it

Goal is to do something that needs no babysitting and still works in most cases.
If we add some package or change to a fork we need to add/change it at the tracker - otherwise we need to add a pkg_id. 
In most cases all that is needed is adding a mapping to the package for LibreELEC. Basically a 3 click action. Not ideal but trusting just the name is not cutting it. Without mapping you roll the dice. Btw github packages will work even we don't add anything.

antonlacon commented:
Handling of release-monitoring.org's package listing should be documented somewhere. Most notably, how to add/remove package, but maybe something else. LIbevent should be removed form the listing if a sample helps.

antonlacon commented on tools/update-scan2:
```diff
@@ -0,0 +1,148 @@
+command -v curl >/dev/null 2>&1 || { echo >&2 "please install curl"; exit 1; }
+command -v jq >/dev/null 2>&1 || { echo >&2 "please install jq"; exit 1; }
+
+. config/options
```
`. config/options ""`

heitbaum commented:
Working on how to use API to ignore black_list and pre-release versions. https://github.com/fedora-infra/anitya/issues/1012

heitbaum commented:
@CvH  - 2 commits to merge into the PR please 96b7a1189879dd996fcea11c4ba636d00ce69806 and 11daeddaaf4e39ac9f1f0bd5d32c74b212dd830a
https://github.com/CvH/LibreELEC.tv/pull/17

CvH commented:
superseded by https://github.com/LibreELEC/LibreELEC.tv/pull/5633
